### ISSUE TITLE
SingleSpaceSeparator Check fails on deeply nested string concatenation

### ISSUE THREAD

**Issue by nmancus1** [checkstyle/checkstyle#8943]
Identified at #8920 
Check documentation: https://checkstyle.sourceforge.io/config_whitespace.html#SingleSpaceSeparator

```
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="SingleSpaceSeparator"/>
    </module>
</module>
➜  src cat ~/testing/jdk14u/test/langtools/tools/javac/DeepStringConcat.java
public class DeepStringConcat {
    public static void main(String[] args) {
        if (X.length() != 32001) throw new Error();
    }
    public static final String X =
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+

"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
... hundreds of lines later ...
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+

"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+
"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+"a"+

        "b";
        /**/
}

➜  src /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 ~/testing/jdk14u/test/langtools/tools/javac/DeepStringConcat.java

➜  src java -jar /home/nick/IdeaProjects/checkstyle/target/checkstyle-8.37-SNAPSHOT-all.jar -c config.xml ~/testing/jdk14u/test/langtools/tools/javac/DeepStringConcat.java
Starting audit...
Exception in thread "main" java.lang.Error: Error was thrown while processing /home/nick/testing/jdk14u/test/langtools/tools/javac/DeepStringConcat.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:319)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:408)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: java.lang.StackOverflowError
        at com.puppycrawl.tools.checkstyle.checks.whitespace.SingleSpaceSeparatorCheck.isTextSeparatedCorrectlyFromPrevious(SingleSpaceSeparatorCheck.java:219)
        at com.puppycrawl.tools.checkstyle.checks.whitespace.SingleSpaceSeparatorCheck.visitEachToken(SingleSpaceSeparatorCheck.java:189)
        at com.puppycrawl.tools.checkstyle.checks.whitespace.SingleSpaceSeparatorCheck.visitEachToken(SingleSpaceSeparatorCheck.java:194)
        at 
... hundreds of lines later ...
        at com.puppycrawl.tools.checkstyle.checks.whitespace.SingleSpaceSeparatorCheck.visitEachToken(SingleSpaceSeparatorCheck.java:194)
        at com.puppycrawl.tools.checkstyle.checks.whitespace.SingleSpaceSeparatorCheck.visitEachToken(SingleSpaceSeparatorCheck.java:194)

➜  src 

```
I'm not sure I would expect Checkstyle to parse this, but it is compilable.

**Comment by romani** [comment]
all our implementations are prone to stackoverflow for very very thousand-nested code. But lets try to fix this.

**Comment by nmancus1** [comment]
Once this issue is fixed, we can remove the "DeepStringConcat.java" `BeforeExecutionExclusionFileFilter` module from `openjdk14-excluded.files'.

**Comment by rnveach** [comment]
This seems like an easy fix.

https://github.com/checkstyle/checkstyle/blob/b23b15328f89d798d15a2e1a17fa7e7e00039d45/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SingleSpaceSeparatorCheck.java#L177
Either we:
1) Have this check do a `visitToken` on every single token, or
2) We basically re-implement the inline recursive "visit every token logic" from `TreeWalker`

**Comment by anhminhtran235** [comment]
I am on it

**Comment by anhminhtran235** [comment]
I'm back guys :) The last couple of weeks have been a busy period for me!

Can I work on this issue?

**Comment by anhminhtran235** [comment]
I'm on it

**Comment by pbludov** [comment]
Fixed in #9102.
